tools/ssllatency.bt

```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/bpftrace
+/*
+ * ssllatency	Trace SSL/TLS handshake for OpenSSL.
+ * 		For Linux, uses bpftrace and eBPF.
+ *
+ * ssllatency shows handshake latency stats and distribution.
+ *
+ * Copyright (c) 2021 Tao Xu.
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ *
+ * 17-Dec-2021	Tao Xu	created this.
+ */
+
+BEGIN
+{
+	printf("Tracing SSL/TLS handshake... Hit Ctrl-C to end.\n");
+}
+
+uprobe:libssl:SSL_read,
+uprobe:libssl:SSL_write,
+uprobe:libssl:SSL_do_handshake
+{
+	@start_ssl[tid] = nsecs;
+	@func_ssl[tid] = func; // store for uretprobe
+}
+
+uretprobe:libssl:SSL_read,
+uretprobe:libssl:SSL_write,
+uretprobe:libssl:SSL_do_handshake
+/@start_ssl[tid] != 0/
+{
+	$lat_us = (nsecs - @start_ssl[tid]) / 1000;
+	if ((int8)retval >= 1) {
+		@hist[@func_ssl[tid]] = lhist($lat_us, 0, 1000, 200);
+		@stat[@func_ssl[tid]] = stats($lat_us);
+	} else {
+		@histF[@func_ssl[tid]] = lhist($lat_us, 0, 1000, 200);
+		@statF[@func_ssl[tid]] = stats($lat_us);
+	}
+	delete(@start_ssl[tid]); delete(@func_ssl[tid]);
+}
+
+// need debug symbol for ossl local functions
```

Comment:
Just wondering - what does this comment mean? Do you need some info from DWARF?

Reply:
The uprobes with "rsa_ossl" prefix are static functions in libcrypto.so that needs DWARF debug info. So *openssl-libs-debuginfo* should be installed for those uprobes in CentOS/Fedora , and *libssl-dev-dbgsym* in Ubuntu.

The rest uprobes in the same *action block* are fine without DWARF. bpftrace allows wildcard to match multiple attach points, and I tested run with no error if no DWARF package installed. I think it shall be fine as far as there's at least one matching uprobe found.